_posts/2022-02-07-sparse-models.md

```diff
@@ -0,0 +1,161 @@
+---
+title: Speeding up Inference with the Lottery Ticket Hypothesis
+date: 2022-02-07
+tags: []
+categories: [Machine Learning]
+image: assets/images/demo1.jpg
+layout: post
+authors: [ojus1]
+---
+
+<script src='https://cdn.plot.ly/plotly-2.8.3.min.js'></script>
+
+The most fancy tool in the Modern Machine Learning toolbox are Neural Networks (NNs), 
+especially `Deep` ones. NNs usually have much more number of parameters than the 
+number of datapoints, and hence are `overparameterized` models. Are all of the parameters 
+needed and perform a useful function in the model? 
+
+The `Lottery Ticket Hypothesis` (LTH) [1] states that for a `reasonably-sized` NN, there 
+exists at least one sub-network (i.e., an NN with some of the parameters/weights
+removed) when trained from scratch that is at least as performant as the full network.
+
+The Lottery Ticket Hypothesis is named as such because of the reasoning that
+there can exist millions of sub-networks in even a relatively small NN; finding
+these well-performing sub-networks (or `winning tickets`) amongst all the possible 
+networks (lottery tickets) is akin to a winning a lottery.
+
+## Sparse Models? Why?
+
+Having sparse models have direct consequences for deployment purposes:
+
+1. If most of the weights are zeros, we can store the weights in CSR/CSC format, and 
+only store the non-zero indices; leading to a very small disk footprint.
+
+2. Matrix Multiplications and other operations for zero-ed elements can be ignored,
+leading to a sharp decline in the number of Floating Point Operations during inference.
+
+3. Performance in terms of robustness to noise and accuracy can acutally `increase` when 
+performing sparsification, as shown by [4] for Automatic Speech Recognition.
+
+## The Search of Winning Tickets
+
+Finding winning tickets reliably is called the `Ticket Search` problem. This is
+usually done by using some sort of weight pruning (zero-ing weights) which are
+less `important`. Each pruning method has its own way of measuring `importance`.
+This pruning is done in an iterative fashion, the procedure of Iterative Magnitude 
+Pruning (IMP) is as follows:
+1. Initialize a network `M` and store its initial weights.
+2. Train the network `M` on the dataset.
+3. Prune `p` number of weights from `M`.
+4. Restore `M` to its initial weight values, except for the pruned weights.
+5. Repeat steps 2-4 `k` times.
+
```

Comment:
Can you help me understand this. Say M = 100 and p = 2. So you initialize a network of 100 nodes and train on a dataset. Then you prune 2 nodes i.e. make them zero. Then in step 4, all the 98 notes are restored to the initial weights except the last 2 nodes which are pruned anyway. Is there some algorithm to chose which weights to prune ? I'm missing something here

Reply:
?